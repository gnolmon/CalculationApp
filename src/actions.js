import {NUMBER_PRESS, OPERATOR_PRESS} from "./constants";

export const buttonPress = (value) => ({
    type: NUMBER_PRESS,
    payload: value
})

export const operatorPress = (value) => ({
    type: OPERATOR_PRESS,
    payload: value
})
