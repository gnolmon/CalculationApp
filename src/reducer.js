import {NUMBER_PRESS, OPERATOR_PRESS} from "./constants";
import {operatorTitles} from "./constants";

const initialState = {
    resultText: "",
    isEqualPress: false
}

function checkIncludeOperator(array) {
    let result = false
    operatorTitles.forEach(item => {
        if (array.includes(item)) {
            result = true
        }
    })
    return result
}

const operatorAtLast = /.[\+\-\*\/]$/g

export default function (state = initialState, action) {
    switch (action.type) {
        case NUMBER_PRESS:
            if (action.payload == 'C') {
                return { resultText: '', isEqualPress: false}
            }
            return { resultText: state.isEqualPress ? action.payload : state.resultText.toString() + action.payload, isEqualPress: false}
        case OPERATOR_PRESS:
            const charArr = state.resultText.toString().split('')

            if (charArr.length == 0) return state

            if (action.payload == '=') {
                if (operatorAtLast.test(state.resultText)) {
                    return {resultText: eval(state.resultText.slice(0, charArr.length - 1)), isEqualPress: checkIncludeOperator(charArr) }
                }
                return {resultText: eval(state.resultText), isEqualPress: checkIncludeOperator(charArr) }
            }

            if (operatorAtLast.test(state.resultText)) {
                charArr[charArr.length - 1] = action.payload
                return { ...state, resultText: charArr.join('')}
            }

            return {resultText: state.resultText + action.payload, isEqualPress: false}
        default:
            return state;
    }
}
