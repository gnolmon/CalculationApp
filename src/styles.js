import {StyleSheet} from "react-native";

export const styles = StyleSheet.create({
    rootContainer: {
        flex: 1,
    },
    displayContainer: {
        flex: 5,
        backgroundColor: '#fff',
        flexDirection: 'column-reverse',
    },
    displayText: {
        color: 'black',
        fontSize: 64,
        textAlign: 'right',
        padding: 15
    },
    inputContainer: {
        flex: 5,
        marginRight: 8,
        marginLeft: 8,
        marginBottom: 20,
        flexDirection: 'row',
        justifyContent: 'space-between'
    },
    numberContainer: {
        flex: 3,
        flexDirection: 'column',
        justifyContent: 'space-around'
    },
    operatorContainer: {
        flex: 1
    },
    inputRow: {
        flex: 1,
        flexDirection: 'row'
    },
    inputButton: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
        margin: 8,
        borderRadius: 8,
        backgroundColor: '#d6d6d6'
    },
    inputButtonHighlighted: {
        backgroundColor: '#193441'
    },
    inputButtonText: {
        fontSize: 34,
        color: 'black'
    },
});
