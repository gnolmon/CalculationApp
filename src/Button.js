import React from 'react'
import {styles} from "./styles";
import {Text, TouchableOpacity} from "react-native";

function Button({value, onPress}) {
    return (
        <TouchableOpacity style={styles.inputButton} onPress={onPress}>
            <Text style={styles.inputButtonText}>{value}</Text>
        </TouchableOpacity>
    )
}

export default React.memo(Button)
