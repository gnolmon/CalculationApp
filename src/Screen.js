import React from 'react';
import { Text, View, TouchableHighlight, SafeAreaView} from 'react-native';
import {styles} from "./styles";
import Button from "./Button";
import {buttonTitles, operatorTitles} from "./constants";
import {operatorPress, buttonPress} from './actions';
import {connect} from 'react-redux';
import { bindActionCreators } from 'redux'
class Screen extends  React.Component {
    constructor(props){
        super(props);
    }

    render() {
        return (
            <SafeAreaView style={styles.rootContainer}>
                <View style={styles.displayContainer}>
                    <Text style={styles.displayText}>{this.props.calculation.resultText}</Text>
                </View>
                <View style={styles.inputContainer}>
                    <View style={styles.numberContainer}>
                        {buttonTitles.map((values, index) => {
                            return (
                                <View style={styles.inputRow} key={`number${index}`}>
                                    {values.map((value, idx) => {
                                        return (
                                            <Button value={value} onPress={() => this.props.buttonPress(value)} key={`btnNumber${idx}`}/>
                                        )
                                    })}
                                </View>
                            );
                        })}
                    </View>
                    <View style={styles.operatorContainer}>
                        {operatorTitles.map((value, index) => {
                            return (
                                <View style={styles.inputRow} key={`operator${index}`}>
                                    <Button value={value} onPress={() => this.props.operatorPress(value)} key={`btnOperator${index}`}/>
                                </View>
                            );
                        })}
                    </View>
                </View>
            </SafeAreaView>
        );
    }
}

export const mapStateToProps = state => ({
    calculation: state
});

export const mapDispatchToProps = dispatch => {
    return bindActionCreators({operatorPress, buttonPress}, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(Screen);
