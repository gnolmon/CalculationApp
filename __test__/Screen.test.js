import React from 'react';
import renderer from 'react-test-renderer';

import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store';
import { mapStateToProps, mapDispatchToProps } from '../src/Screen';
import Screen from '../src/Screen';

const mockStore = configureStore();

const initialState = {
    resultText: "",
    isEqualPress: false
}

describe('<Screen />', () => {
    let wrapper, store;

    beforeEach(() => {
        const initialState = {
            lastRolledNumber: 1
        };
        store = mockStore(initialState);
        wrapper = shallow(<Screen store={store} />);
    });

    it('renders correctly', () => {
        const tree = renderer.create(<Screen store={store}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should call action NUMBER_PRESS when click number', () => {
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).buttonPress(1);
        expect(dispatch.mock.calls[0][0]).toEqual({ type: 'NUMBER_PRESS', payload: 1});
    });

    it('should call action OPERATOR_PRESS when click operator', () => {
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).operatorPress("+");
        expect(dispatch.mock.calls[0][0]).toEqual({ type: 'OPERATOR_PRESS', payload: "+"});
    });

});

