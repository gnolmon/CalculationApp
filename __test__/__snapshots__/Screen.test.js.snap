// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Screen /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 5,
        "flexDirection": "column-reverse",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 64,
          "padding": 15,
          "textAlign": "right",
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 20,
        "marginLeft": 8,
        "marginRight": 8,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 3,
          "flexDirection": "column",
          "justifyContent": "space-around",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            7
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            8
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            9
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            4
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            5
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            6
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            1
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            2
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            3
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            C
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            .
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            /
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            *
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            +
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            -
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#d6d6d6",
              "borderRadius": 8,
              "flex": 1,
              "justifyContent": "center",
              "margin": 8,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 34,
              }
            }
          >
            =
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
